File: LeetCode/36. Valid Sudoku/36.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxes = [set() for _ in range(9)]
        
        for i in range(9):
            for j in range(9):
                if board[i][j] != '.':
                    num = board[i][j]
                    boxIndex = (i // 3) * 3 + j // 3
                    
                    # Check if this value is already in the row, column, or box
                    if num in rows[i] or num in cols[j] or num in boxes[boxIndex]:
                        return False
                    
                    # Add the number to the current row, column, and box
                    rows[i].add(num)
                    cols[j].add(num)
                    boxes[boxIndex].add(num)
        
        return True
    

    '''
    class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        seen = set()
        for i in range(9):
            for j in range(9):
                current = board[i][j]
                if current != ".":
                    if (i, current) in seen or (current,j) in seen or (i//3,j//3,current) in seen:
                        return False
                    seen.add((i,current))
                    seen.add((current,j))
                    seen.add((i//3,j//3,current))
                
        return True
    '''

File: LeetCode/36. Valid Sudoku/36.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<set<int>> rows(9), cols(9), boxes(9);
        
        for(int i = 0; i < 9; i++) {
            for(int j = 0; j < 9; j++) {
                if(board[i][j] != '.') {
                    int num = board[i][j] - '0';
                    int boxIndex = (i / 3) * 3 + j / 3;
                    
                    // Check if this value is already in the row, column, or box
                    if(rows[i].count(num) || cols[j].count(num) || boxes[boxIndex].count(num))
                        return false;
                    
                    // Add the number to the current row, column, and box
                    rows[i].insert(num);
                    cols[j].insert(num);
                    boxes[boxIndex].insert(num);
                }
            }
        }
        
        return true;
    }
};

/*
class Solution
{
public:
    bool isValidSudoku(vector<vector<char> > &board)
    {
        int used1[9][9] = {0}, used2[9][9] = {0}, used3[9][9] = {0};
        
        for(int i = 0; i < board.size(); ++ i)
            for(int j = 0; j < board[i].size(); ++ j)
                if(board[i][j] != '.')
                {
                    int num = board[i][j] - '0' - 1, k = i / 3 * 3 + j / 3;
                    if(used1[i][num] || used2[j][num] || used3[k][num])
                        return false;
                    used1[i][num] = used2[j][num] = used3[k][num] = 1;
                }
        
        return true;
    }
};
*/